### DISCUSSION TITLE
tensor_mechanics_J2plasticity

### DISCUSSION THREAD

**Discussion by rh201** [idaholab/moose#23426] [Q&A Modules: Solid mechanics] [answered]
Hi MOOSEers,

I'm trying to do benchmark with tensor_mechanics_J2plasticity based on  [[modules/tensor_mechanics/test/tests/j2_plasticity/tensor_mechanics_j2plasticity.i](https://github.com/idaholab/moose/blob/afc31dc62b82174f8d95a76cf1b46c18bf44438d/modules/tensor_mechanics/test/tests/j2_plasticity/tensor_mechanics_j2plasticity.i)](url). A monolithic tensile loading is applied on one surface of the cubic element, but when I check the load-displacement curve, the force does not keep increasing, instead, it decreases and shows softening after the peak load.  I also tried with [IsotropicPlasticityStressUpdate] and [ComputeMultipleInelasticStress], and the results are similar, which is confusing. Are these results correct?  Thanks.

**Comment by rh201** [comment]
I'm using the input file as below, with [IsotropicPlasticityStressUpdate] and [ComputeMultipleInelasticStress] for plasticity. When the external loading is relatively small, the load-displacement curve looks good, with the bi-linear region for elasticity and linear hardening, respectively. But when the loading further increases, the curve behaves wired: the hardening region becomes curved, and further the curve goes down. How to explain this? Thanks.

`[GlobalParams]
  displacements = 'disp_x disp_y disp_z'
[]

[Mesh]
  [generated_mesh]
    type = GeneratedMeshGenerator
    elem_type = HEX8
    dim = 3
    nx = 1
    ny = 1
    nz = 1
    xmin = 0.0
    xmax = 1.0
    ymin = 0.0
    ymax = 1.0
    zmin = 0.0
    zmax = 1.0
  []
  [cnode]
    type = ExtraNodesetGenerator
    coord = '0.0 0.0 0.0'
    new_boundary = corner_point1
    input = generated_mesh
  []
  [snode]
    type = ExtraNodesetGenerator
    coord = '1.0 0.0 0.0'
    new_boundary = corner_point2
    input = cnode
  []
[]

[Variables]
  [disp_x]
    order = FIRST
    family = LAGRANGE
  []
  [disp_y]
    order = FIRST
    family = LAGRANGE
  []
  [disp_z]
    order = FIRST
    family = LAGRANGE
  []
[]

[AuxVariables]
  [fx]
  []
  [fy]
  []
  [fz]
  []
 []

[Kernels]
  [./TensorMechanics]
    displacements = 'disp_x disp_y disp_z'
    use_displaced_mesh = true
    use_automatic_differentiation = true
    save_in = 'fx fy fz'
  [../]
[]

[BCs]
  [ydisp]
    type = FunctionDirichletBC
    variable = disp_y
    boundary = top
    function = 't'
  []
  [yfix]
    type = DirichletBC
    variable = disp_y
    boundary = bottom
    value = 0
  []
  [./corner1_x]
    type = DirichletBC
    variable = disp_x
    boundary = corner_point1
    value = 0.0
  [../]
  [./corner1_y]
    type = DirichletBC
    variable = disp_y
    boundary = corner_point1
    value = 0.0
  [../]
  [./corner1_z]
    type = DirichletBC
    variable = disp_z
    boundary = corner_point1
    value = 0.0
  [../]
  [./corner2_y]
    type = DirichletBC
    variable = disp_y
    boundary = corner_point2
    value = 0.0
  [../]
  [./corner2_z]
    type = DirichletBC
    variable = disp_z
    boundary = corner_point2
    value = 0.0
  [../]
[]

[Materials]
  [strain]
    type = ADComputeFiniteStrain  #ComputeSmallStrain
  []
  [elasticity_tensor]
    type = ADComputeIsotropicElasticityTensor
    youngs_modulus = 68.8e3
    poissons_ratio = 0.33
  []
  [./isotropic_plasticity]
    type = ADIsotropicPlasticityStressUpdate
    yield_stress = 320
    hardening_constant = 650
    relative_tolerance = 1e-8
    absolute_tolerance = 1e-11
    max_inelastic_increment = 0.0001
  [../]
  [./radial_return_stress]
    type = ADComputeMultipleInelasticStress
    # tangent_operator = elastic
    inelastic_models = 'isotropic_plasticity'
  [../]
[]

[Postprocessors]
  [fx]
    type = NodalSum
    variable = fx
    boundary = top
  []
  [fy]
    type = NodalSum
    variable = fy
    boundary = top
  []
  [fz]
    type = NodalSum
    variable = fz
    boundary = top
  []
[]

[Executioner]
  type = Transient
  solve_type = PJFNK
  petsc_options_iname = '-pc_type'
  petsc_options_value = 'lu'

  dt=0.001
  dtmax=0.1
  dtmin=0.001
  end_time=10.0

  nl_abs_tol = 1e-10
[]

[Outputs]
  exodus = true
  print_linear_residuals = false
  [csv]
    type = CSV
    delimiter = ' '
    file_base = 'f'
  []
[]
`

**Comment by hugary1995** [comment, marked as answer]
The load-displacement curve would be equivalent to an engineering stress-engineering strain relation. You'll want to plot true stress vs the logorithmic strain.

**Reply by rh201** [reply]
Thanks  @hugary1995 . How to plot the true stress-logorithmic strain curve for this case?

**Reply by hugary1995** [reply]
I setup the input file (see below at the end) for you. Take a look at the postprocessors, they should be self-explanatory. The stress-strain curves are shown below.

![stress](https://user-images.githubusercontent.com/31555204/219378747-3479b5fa-c80d-497c-b45b-53d537e3cc01.png)

<details>
  <summary>input file</summary>

```python
[GlobalParams]
  displacements = 'disp_x disp_y disp_z'
[]
[Mesh]
  [generated_mesh]
    type = GeneratedMeshGenerator
    elem_type = HEX8
    dim = 3
    nx = 1
    ny = 1
    nz = 1
    xmin = 0.0
    xmax = 1.0
    ymin = 0.0
    ymax = 1.0
    zmin = 0.0
    zmax = 1.0
  []
[]
[Variables]
  [disp_x]
  []
  [disp_y]
  []
  [disp_z]
  []
[]
[AuxVariables]
  [fx]
  []
  [fy]
  []
  [fz]
  []
[]
[Kernels]
  [TensorMechanics]
    displacements = 'disp_x disp_y disp_z'
    use_displaced_mesh = true
    use_automatic_differentiation = true
    save_in = 'fx fy fz'
  []
[]
[BCs]
  [xfix]
    type = DirichletBC
    variable = disp_x
    value = 0
    boundary = left
  []
  [yfix]
    type = DirichletBC
    variable = disp_y
    value = 0
    boundary = bottom
  []
  [zfix]
    type = DirichletBC
    variable = disp_z
    value = 0
    boundary = back
  []
  [pull]
    type = FunctionDirichletBC
    variable = disp_z
    function = t
    boundary = front
  []
[]
[Materials]
  [strain]
    type = ADComputeFiniteStrain
  []
  [elasticity_tensor]
    type = ADComputeIsotropicElasticityTensor
    youngs_modulus = 68.8e3
    poissons_ratio = 0.33
  []
  [isotropic_plasticity]
    type = ADIsotropicPlasticityStressUpdate
    yield_stress = 320
    hardening_constant = 650
    relative_tolerance = 1e-8
    absolute_tolerance = 1e-11
    max_inelastic_increment = 0.0001
  []
  [radial_return_stress]
    type = ADComputeMultipleInelasticStress
    inelastic_models = 'isotropic_plasticity'
  []
[]
[Postprocessors]
  [F]
    type = NodalSum
    variable = fz
    boundary = front
    execute_on = 'INITIAL TIMESTEP_END'
  []
  [A0]
    type = AreaPostprocessor
    boundary = front
    execute_on = 'INITIAL TIMESTEP_END'
  []
  [A]
    type = AreaPostprocessor
    boundary = front
    use_displaced_mesh = true
    execute_on = 'INITIAL TIMESTEP_END'
  []
  [engineering_stress]
    type = ParsedPostprocessor
    function = 'F/A0'
    pp_names = 'F A0'
    execute_on = 'INITIAL TIMESTEP_END'
  []
  [true_stress]
    type = ParsedPostprocessor
    function = 'F/A'
    pp_names = 'F A'
    execute_on = 'INITIAL TIMESTEP_END'
  []
  [engineering_strain]
    type = FunctionValuePostprocessor
    function = 't'
    execute_on = 'INITIAL TIMESTEP_END'
  []
  [log_strain]
    type = FunctionValuePostprocessor
    function = 'log(1+t)'
    execute_on = 'INITIAL TIMESTEP_END'
  []
[]
[Executioner]
  type = Transient
  solve_type = NEWTON
  petsc_options_iname = '-pc_type'
  petsc_options_value = 'lu'
  automatic_scaling = true
  line_search = none
  dt = 0.005
  end_time = 2
  nl_rel_tol = 1e-08
  nl_abs_tol = 1e-10
[]
[Outputs]
  exodus = true
  csv = true
[]
```
</details>

**Reply by hugary1995** [reply]
As a quick sanity check: for the linear hardening law
true stress = yield stress + hardening modulus * log plastic strain

When strain is large, most portion of the total strain is plastic, hence
true stress ~= yield stress + hardening modulus * log total strain

So the true stress @ log total strain = 1 is
F/A ~= 320 + 650 * 1 = 970 MPa
The true stress we get numerically @ log strain = 1 is approximately 960 MPa.

**Reply by rh201** [reply]
Thanks!

**Reply by rh201** [reply]
@hugary1995  I used the input file you provided and got the data (below) for log strain and true stress, but the curve plotted out of it is still not a straight line, maybe I took something wrong, can you help me have a quick check? Thanks.
[input_out.csv](https://github.com/idaholab/moose/files/10759886/input_out.csv)

**Reply by hugary1995** [reply]
This csv is identical to mine, which is what I used to generate the above figure.

**Reply by rh201** [reply]
Thank you ! @hugary1995
